File: Concepts/LinkedLists/Java/DoublyLinkedList.java
import java.util.Scanner;
public class DoublyLinkedList {

    DoublyLinkedList() {
        Scanner sc = new Scanner(System.in);
        while (true) {
            System.out.println("\n----------------------------");
            System.out.println("1. Insert At Head");
            System.out.println("2. Insert At Tail");
            System.out.println("3. Delete At Head");
            System.out.println("4. Delete At Tail");
            System.out.println("5. Traverse");
            System.out.println("6. Reverse Traverse");
            System.out.println("(Press any key to exit)");
            System.out.println("----------------------------");
            System.out.println("Enter an option: ");
            switch (sc.nextInt()) {
                case 1:
                    System.out.println("Enter the value: ");
                    insertAtHead(sc.nextInt());
                    break;
                    case 2:
                    System.out.println("Enter the value: ");
                    insertAtTail(sc.nextInt());
                    break;
                case 3:
                    deleteAtHead();
                    break;
                case 4:
                    deleteAtTail();
                    break;
                case 5:
                    print();
                    break;
                case 6:
                    revPrint();
                    break;
                default:
                    return;
            }
        }
    }
    static class Node {
        Node prev;
        int data;
        Node next;

        Node(int item) {
            prev = null;
            data = item;
            next = null;
        }
    }

    Node head;

    void insertAtHead(int item) {
        Node newNode = new Node(item);
        if (head != null) {
            head.prev = newNode;
        }
        newNode.next = head;
        head = newNode;
        
    }

    void insertAtTail(int item) {
        Node newNode = new Node(item);
        if (head == null) {
            head = newNode;
            return;
        }
        Node currNode = head;
        while (currNode.next != null)
            currNode = currNode.next;
        newNode.prev = currNode;
        currNode.next = newNode;
    }
    
    void deleteAtHead() {
        Node temp = head;
        head = head.next;
        head.prev = null;
        temp = null;
    }
    
    void deleteAtTail() {
        Node currNode = head;
        while (currNode.next.next != null)
            currNode = currNode.next;
        currNode.next.prev = null;
        currNode.next = null;
    }
    
    void print() {
        Node currNode = head;
        while (currNode != null) {
            System.out.print(currNode.data + " -> ");
            currNode = currNode.next;
        }
        System.out.print("null\n");
    }

    void revPrint() {
        Node currNode = head;
        while (currNode.next != null) {
            currNode = currNode.next;
        }
        while (currNode != null) {
            System.out.print(currNode.data + " -> ");
            currNode = currNode.prev;
        }
        System.out.print("null\n");
    }


    public static void main(String[] args) {
        DoublyLinkedList foo = new DoublyLinkedList();
    }
}

File: Concepts/LinkedLists/Python/DoublyLinkedList.py
class Node:
    def __init__(self, val: int):
        self.prev = None
        self.item = val
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        while True:
            print("\n----------------------------")
            print("1. Insert At Head")
            print("2. Insert At Tail")
            print("3. Delete At Head")
            print("4. Delete At Tail")
            print("5. Traverse")
            print("6. Reverse Traverse")
            print("(Press any key to exit)")
            print("----------------------------")
            print("Enter an option: ")
            match (int(input())):
                case 1:
                    self.insert_at_head(int(input("Enter the value: ")))
                case 2:
                    self.insert_at_tail(int(input("Enter the value: ")))
                case 3:
                    self.delete_at_head()
                case 4:
                    self.delete_at_tail()
                case 5:
                    self.printList()
                case 6:
                    self.printRevList()
                case _:
                    break
    
    def insert_at_head(self, val: int):
        newNode = Node(val)
        newNode.next = self.head
        if self.head != None:
            self.head.prev = newNode
        self.head = newNode
    
    def insert_at_tail(self, val: int):
        newNode = Node(val)
        if self.head == None:
            head = newNode
        currNode = self.head
        while currNode.next != None:
            currNode = currNode.next
        newNode.prev = currNode
        currNode.next = newNode

    def delete_at_tail(self):
        currNode = self.head
        while currNode.next.next != None:
            currNode = currNode.next
        currNode.next.prev = None
        currNode.next = None
    
    def delete_at_head(self):
        self.head = self.head.next
        self.head.prev = None
    
    def printList(self):
        if self.head == None:
            print("List is empty")
        currNode = self.head
        while currNode != None:
            print(currNode.item, " -> ", end="")
            currNode = currNode.next
        print("null")
        
    def printRevList(self):
        if self.head == None:
            print("List is empty")
        currNode = self.head
        while currNode.next != None:
            currNode = currNode.next
        while currNode != None:
            print(currNode.item, " -> ", end="")
            currNode = currNode.prev
        print("null")

ll = DoublyLinkedList()
